curp/src/client.rs

```diff
@@ -1,30 +1,198 @@
         ReadState as PbReadState, SyncResult, WaitSyncedRequest,
     },
     LogIndex, ServerId,
 };
 
+/// Client builder
+#[derive(Debug)]
+pub struct Builder<C> {
+    /// Server addresses
+    addrs: Vec<String>,
+    /// Server addresses and ids (used in an inner client)
+    all_members: HashMap<ServerId, String>,
+    /// Local server id (used in an inner client)
+    local_server_id: Option<ServerId>,
+    /// Client timeout
+    timeout: Option<ClientTimeout>,
+    /// Phantom data
+    phantom: PhantomData<C>,
+}
+
+impl<C> Default for Builder<C> {
+    #[inline]
+    #[must_use]
+    fn default() -> Self {
+        Self {
+            addrs: Vec::new(),
+            all_members: HashMap::new(),
+            local_server_id: None,
+            timeout: None,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<C> Builder<C>
+where
+    C: Command + 'static,
+{
+    /// Create new client builder
+    #[inline]
+    #[must_use]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set addresses of servers
+    /// # Panics
+    /// Panics when set `all_members` before
+    #[inline]
+    #[must_use]
+    pub fn addrs(self, addrs: Vec<String>) -> Self {
+        Self { addrs, ..self }
+    }
+
+    /// Set server id and all addresses of servers
+    /// # Panics
+    /// Panics when set `addrs` before
+    #[inline]
+    #[must_use]
+    pub fn all_members(self, all_members: HashMap<ServerId, String>) -> Self {
+        Self {
+            all_members,
+            ..self
+        }
+    }
+
+    /// Set client timeout
+    #[inline]
+    #[must_use]
+    pub fn timeout(self, timeout: ClientTimeout) -> Self {
+        Self {
+            timeout: Some(timeout),
+            ..self
+        }
+    }
+
+    /// Set local server id. (only used in an inner client)
+    #[inline]
+    #[must_use]
+    pub fn local_server_id(self, local_server_id: ServerId) -> Self {
+        Self {
+            local_server_id: Some(local_server_id),
+            ..self
+        }
+    }
+
+    /// Build client
+    /// # Errors
+    /// Return error when meet rpc error or missing some arguments
+    #[inline]
+    pub async fn build(self) -> Result<Client<C>, ClientBuildError> {
+        let Some(timeout) = self.timeout else {
+            return Err(ClientBuildError::invalid_aurguments("timeout is required"));
+        };
+        match (self.addrs.is_empty(), self.all_members.is_empty()) {
+            (false, true) => {
+                Self::build_from_addrs(timeout, self.local_server_id, self.addrs).await
+            }
+            (true, false) => {
+                Self::build_from_all_members(
+                    timeout,
+                    self.local_server_id,
+                    self.all_members,
+                    None,
+                    0,
+                )
+                .await
+            }
+            (true, true) => Err(ClientBuildError::invalid_aurguments(
```

Comment:
Why should we return Error when the user provides the `addr` and the `all_members`? It's okay to ignore one of them and build Client with the other.

Reply:
Fixed with other implementation